src/lib/tab_groups.ts

```diff
@@ -0,0 +1,330 @@
+            const ids = tabs.map(tab => tab.id)
+            await browser.tabs.show(ids)
+            return clearTabTgroup(ids)
+        })
+    ])
+}
+
+/**
+ * Set the tab's tab group to the window's active tab group if there is one.
+ *
+ * Do nothing if the tab is already associated with a tab group.
+ *
+ * @param tab The tab that was just created.
+ *
+ */
+export async function tgroupHandleTabCreated(tab: browser.tabs.Tab) {
+    const windowGroup = await windowTgroup() as string
+
+    if (windowGroup) {
+        const tabGroup = await tabTgroup(tab.id)
+        if (!tabGroup) {
+            return setTabTgroup(windowGroup, tab.id)
+        }
+    }
+}
+
+/**
+ * Set the tab's tab group to the window's active tab group if there is one.
+ *
+ * @param tabId The id of the tab that was just attached to this window.
+ *
+ */
+export async function tgroupHandleTabAttached(tabId: number, _attachInfo: any) {
+    // NOTE this doesn't need to worry about a tab on another window previously
+    // being pinned because tabs become unpinned when you move them between
+    // windows
+    const windowGroup = await windowTgroup() as string
+    if (windowGroup) {
+        return setTabTgroup(windowGroup, tabId)
+    }
+}
+
+// TODO where is active info's type?
+/**
+ * Handle tab activation, possibly switching tab groups.
+ *
+ * If the new tab is from a different tab group, set it as the new tab group,
+ * show all its tabs, and hide all tabs from the old tab group.
+ */
+export async function tgroupHandleTabActivated(activeInfo) {
+    const windowGroup = await windowTgroup(activeInfo.windowId) as string
+    const tabGroup = await tabTgroup(activeInfo.tab) as string
+    const promises = []
+    if ((windowGroup && tabGroup) && (windowGroup != tabGroup)) {
+        await setWindowTgroup(tabGroup)
+
+        promises.push(tgroupRun(tabGroup).then(tabs => browserBg.tabs.show(tabs.map(tab => tab.id))))
+        promises.push(tgroupRun(tabGroup, true).then(tabs => browserBg.tabs.hide(tabs.map(tab => tab.id))))
+    }
+    return Promise.all(promises)
+}
+
+// TODO changeInfo type
+/**
+ * Set or clear a tab's group if it was pinned or unpinned respectively.
+ *
+ * @param tabId The id of the tab that was just updated.
+ *
+ */
+export async function tgroupHandleTabUpdated(tabId: number, changeInfo) {
+    if (changeInfo.pinned !== undefined) {
+        const windowGroup = await windowTgroup() as string
+        if (windowGroup) {
+            if (changeInfo.pinned) {
+                return clearTabTgroup(tabId)
+            } else {
+                return setTabTgroup(windowGroup, tabId)
+            }
+        }
+    }
+}
+
+// TODO removeInfo type
+/**
+ * Handle the last tab in a tab group being closed.
+ *
+ * Clear its information.
+ *
+ */
+export async function tgroupHandleTabRemoved(_tabId: number, removeInfo) {
+    if (!removeInfo.isWindowClosing) {
+        const windowGroup = await windowTgroup() as string
```

Comment:
```suggestion
        const windowGroup = await windowTgroup()
```
If you add the cast to windowTgroup.